### PR TITLE
fix(cli): Guard serve fast-path bundle closure

### PR DESCRIPTION
**Pull request by doudouOUC**
## What this PR does

This PR adds a dedicated bundle-level guard for the `qwen serve` fast path so the pre-listen static bundle closure is checked against ACP bridge runtime modules, core shell runtime, and the heavy vendor packages that previously entered through shared chunks. It also keeps the lightweight source-level guard for the request-helper import boundary and moves the expensive real-bundle assertion out of the CLI unit test path into an explicit CI check.

## Why it's needed

The daemon startup optimization depends on keeping the listener path free of eager ACP runtime and transitive core/vendor imports. Source-level import checks alone can miss esbuild shared chunks and hashed output names, so this adds a metafile-based reachability check that follows static output imports while allowing dynamic imports used by the runtime loader.

### How to verify

Run `npm run test:scripts -- scripts/tests/serve-fast-path-bundle-check.test.js` and confirm the synthetic metafile coverage catches static forbidden runtime imports, allows dynamic imports, reports the vendor package sentinels, and prints the static import path diagnostic. Run `cd packages/cli && npx vitest run src/serve/fast-path.test.ts` and confirm the source fast-path tests pass. Run `npm run check:serve-fast-path-bundle` and confirm the real bundled serve fast-path closure passes. Run `npm run build && npm run typecheck` and confirm the workspace still builds and typechecks.

### Evidence (Before & After)

N/A. This is a non-UI guard and import-boundary hardening change.

### Tested on

|     OS     | Status |
| :--------: | :----: |
|  🍏 macOS  | ✅ tested |
| 🪟 Windows | ⚠️ not tested |
|  🐧 Linux  | ⚠️ not tested |

### Environment (optional)

macOS local checkout, Node v26.0.0, npm 11.12.1.

## Risk & Scope

- Main risk or tradeoff: The new CI check performs a CLI-only build and bundle on the Ubuntu CI leg, which adds CI time but gives direct coverage for the bundled daemon fast path. It intentionally does not repeat this guard in the macOS/Windows merge-queue jobs because the bundle import graph is platform-independent and the script normalizes paths.
- Not validated / out of scope: This does not change `getCliVersion()` timing, raw HTTP/Express bootstrap behavior, settings async reads, or package `sideEffects` metadata.
- Breaking changes / migration notes: None.

## Linked Issues

Refs #4748

<details>
<summary>中文说明</summary>

## What this PR does

这个 PR 为 `qwen serve` fast path 增加了独立的 bundle-level guard，用来检查 pre-listen static bundle closure 是否包含 ACP bridge runtime、core shell runtime，以及此前通过 shared chunks 进入的重 vendor packages。同时保留 request-helper import 边界的轻量源码级 guard，并把真实 bundle 的重断言从 CLI 单测路径移到显式 CI 检查里。

## Why it's needed

daemon 启动优化依赖 listener 路径不 eager 加载 ACP runtime 和 transitive core/vendor imports。单纯的源码 import 检查会漏掉 esbuild shared chunks 和 hash 后的输出文件名，所以这里新增基于 metafile 的 reachability check：跟踪静态 output imports，同时允许 runtime loader 使用的 dynamic imports。

### How to verify

运行 `npm run test:scripts -- scripts/tests/serve-fast-path-bundle-check.test.js`，确认 synthetic metafile 覆盖可以捕获静态 forbidden runtime imports、允许 dynamic imports、报告 vendor package sentinels，并输出 static import path 诊断。运行 `cd packages/cli && npx vitest run src/serve/fast-path.test.ts`，确认源码 fast-path 测试通过。运行 `npm run check:serve-fast-path-bundle`，确认真实 bundled serve fast-path closure 通过。运行 `npm run build && npm run typecheck`，确认 workspace 仍可 build 和 typecheck。

### Evidence (Before & After)

N/A。这是非 UI 的 guard 和 import-boundary hardening 变更。

### Tested on

|     OS     | Status |
| :--------: | :----: |
|  🍏 macOS  | ✅ tested |
| 🪟 Windows | ⚠️ not tested |
|  🐧 Linux  | ⚠️ not tested |

### Environment (optional)

macOS 本地 checkout，Node v26.0.0，npm 11.12.1。

## Risk & Scope

- Main risk or tradeoff: 新 CI 检查会在 Ubuntu CI leg 上执行 CLI-only build 和 bundle，会增加 CI 时间，但能直接覆盖 bundled daemon fast path。它有意不在 macOS/Windows merge-queue jobs 中重复运行这个 guard，因为 bundle import graph 与平台无关，且脚本已经做了路径归一化。
- Not validated / out of scope: 本 PR 不修改 `getCliVersion()` timing、raw HTTP/Express bootstrap behavior、settings async reads，也不修改 package `sideEffects` metadata。
- Breaking changes / migration notes: 无。

## Linked Issues

Refs #4748

</details>

### CODE REVIEW

**Comment by wenshao** [comment]
## Verification report — PR #5995 (Guard serve fast-path bundle closure)

I checked out this PR head (`a6cc23cca0`), built the real CLI bundle, and ran the new guard end-to-end — including a **mutation test** to prove the guard actually catches a regression, not just passes vacuously. This is a non-UI build/import-boundary PR, so the "real test" here is the real bundle + guard + a deliberately broken bundle, rather than a TUI session.

### TL;DR
**Looks good — no blocking issues.** The new bundle guard is correct, well-tested, and **provably load-bearing**: it passes on the clean bundle and fails with a precise diagnostic the moment a forbidden ACP runtime module is statically pulled into the serve pre-listen closure. The refactor (moving the expensive real-bundle assertion out of the per-platform vitest path into a dedicated ubuntu CI check) is a net improvement and strictly broader in coverage than the test it replaces.

| Check | Result |
|---|---|
| Clean bundle → guard | ✅ PASS (`exit 0`, "closure check passed") |
| **Mutation** (static `import '@qwen-code/acp-bridge/bridge'` into a pre-listen root) → guard | ✅ **FAIL as designed** (`exit 1`, offender + import path reported) |
| Restore → guard | ✅ PASS again (`exit 0`) — clean A→B→A loop |
| Script unit tests (`serve-fast-path-bundle-check.test.js`) | ✅ 12/12 |
| Source guard tests (`fast-path.test.ts`) | ✅ 58/58 |
| `tsc --noEmit` (cli package) | ✅ 0 errors |
| CI wiring (`ci.yml`) | ✅ step added to the ubuntu `test` job, gated on `skip_ci` |
| `eventBus` vitest alias | ✅ justified (cli serve files value/type-import `@qwen-code/acp-bridge/eventBus`; matches existing alias pattern) |

### Setup
- Worktree at PR head `a6cc23cca0`; `npm ci`; `npm run check:serve-fast-path-bundle` (= `build --cli-only` → `DEV=true bundle` → guard). Real 5.2 MB `dist/esbuild.json` metafile. macOS, Node 22.

### The decisive mutation test (A → B → A)

The whole value of a guard is that it fails when it should. I converted the legitimate **dynamic** ACP load into a **static** one by adding `import '@qwen-code/acp-bridge/bridge';` to a pre-listen root (`run-qwen-serve.ts`), rebuilt the real bundle, and re-ran the guard:

```text
A. clean bundle      → Serve fast-path bundle closure check passed.            (exit 0)

B. mutated bundle    → Serve fast-path bundle closure includes pre-listen runtime modules:
   - ACP bridge runtime
     input:  packages/acp-bridge/src/bridge.ts
     output: dist/chunks/chunk-YJNSEL6A.js (179167 bytes)
     static path: dist/chunks/run-qwen-serve-HOEXETRE.js -> dist/chunks/chunk-YJNSEL6A.js
   - ACP compaction engine runtime / ACP bridge client runtime / ACP permission mediator runtime ...
                                                                                  (exit 1)

A'. restored bundle  → Serve fast-path bundle closure check passed.            (exit 0)
```

That single static import pulled the **entire ~179 KB ACP runtime cluster** (bridge → compactionEngine, bridgeClient, permissionMediator, …) into one statically-reachable chunk — exactly the "heavy runtime entering the listener path through a shared chunk" regression this guard exists to block. The metafile-reachability approach (follow static `import`/`require` edges, skip `dynamic-import` + `external`) correctly catches it, while the dynamic-import path stays allowed (covered by the script's unit tests).

### Notes / trade-offs (non-blocking)
- **Coverage moved, and broadened.** The removed in-vitest assertion only checked the `run-qwen-serve` root against 4 forbidden inputs; the new script checks all 3 pre-listen roots against 7 ACP/core-shell runtime modules **plus** 4 heavy vendor packages (glob / chokidar / @iarna/toml / fzf), with import-path diagnostics. Strictly more coverage.
- **Runs on ubuntu only** (the `test` job's static-check leg), vs. the old assertion that ran in every platform's vitest. The bundle import-graph is platform-independent, so ubuntu-only is sufficient, and it removes a ~30 s esbuild run from each of the 3 platforms' unit suites — net CI win (the PR notes the added CLI-only build cost).
- **No longer part of `npm test`.** A bundle-level regression is now caught by `npm run check:serve-fast-path-bundle` / CI rather than a bare `npm test`. The fast source-level guards in `fast-path.test.ts` still run in unit tests, so source-level violations are still caught locally; only the esbuild shared-chunk edge cases require the dedicated check. Deliberate and reasonable.
- The guard fails closed if a root output can't be found ("Could not find bundled outputs …") — a safe default (covered by a test).

### Recommendation
**No changes requested.** The guard does what it claims, is well-tested, and the import boundary holds on the real bundle. Recommend merge once the in-progress CI goes green — at report time the new **"Check serve fast-path bundle closure"** step was still running on the ubuntu leg (it passes locally); mac/windows are the expected named-job skips.

<details>
<summary>中文版（完整对应）</summary>

## 验证报告 — PR #5995（给 serve fast-path bundle closure 加 guard）

我 checkout 了本 PR head（`a6cc23cca0`），构建了真实 CLI bundle，并端到端运行了新 guard —— 包含一次**变异测试（mutation test）**，用来证明 guard 真的能抓到回归，而不是空过。这是个非 UI 的构建/import 边界 PR，所以这里的"真实测试"是真实 bundle + guard + 故意打破的 bundle，而不是 TUI 会话。

### 结论速览
**没问题 —— 无阻塞项。** 新的 bundle guard 正确、测试充分，且**确实承重**：在干净 bundle 上通过，一旦有 forbidden ACP runtime 模块被静态拉进 serve pre-listen closure，它立刻以精确诊断失败。这次重构（把昂贵的真实 bundle 断言从逐平台 vitest 路径移到独立的 ubuntu CI 检查）是净改进，且覆盖范围严格大于它替换掉的那个测试。

| 检查 | 结果 |
|---|---|
| 干净 bundle → guard | ✅ PASS（`exit 0`，"closure check passed"） |
| **变异**（往 pre-listen root 加静态 `import '@qwen-code/acp-bridge/bridge'`）→ guard | ✅ **如期 FAIL**（`exit 1`，报告 offender + import path） |
| 还原 → guard | ✅ 再次 PASS（`exit 0`）—— 完整 A→B→A 闭环 |
| 脚本单测（`serve-fast-path-bundle-check.test.js`） | ✅ 12/12 |
| 源码 guard 测试（`fast-path.test.ts`） | ✅ 58/58 |
| `tsc --noEmit`（cli 包） | ✅ 0 错误 |
| CI 接线（`ci.yml`） | ✅ 步骤加到 ubuntu `test` job，按 `skip_ci` 门控 |
| `eventBus` vitest alias | ✅ 合理（cli serve 文件 value/type 导入 `@qwen-code/acp-bridge/eventBus`；与既有 alias 模式一致） |

### 环境
- worktree 切到 PR head `a6cc23cca0`；`npm ci`；`npm run check:serve-fast-path-bundle`（= `build --cli-only` → `DEV=true bundle` → guard）。真实 5.2 MB `dist/esbuild.json` metafile。macOS，Node 22。

### 决定性的变异测试（A → B → A）

guard 的全部价值在于"该失败时能失败"。我把合法的**动态** ACP 加载改成**静态**（往 pre-listen root `run-qwen-serve.ts` 加 `import '@qwen-code/acp-bridge/bridge';`），重新构建真实 bundle，再跑 guard：

```text
A. 干净 bundle    → Serve fast-path bundle closure check passed.            (exit 0)

B. 变异 bundle    → Serve fast-path bundle closure includes pre-listen runtime modules:
   - ACP bridge runtime
     input:  packages/acp-bridge/src/bridge.ts
     output: dist/chunks/chunk-YJNSEL6A.js (179167 bytes)
     static path: dist/chunks/run-qwen-serve-HOEXETRE.js -> dist/chunks/chunk-YJNSEL6A.js
   - ACP compaction engine runtime / ACP bridge client runtime / ACP permission mediator runtime ...
                                                                            (exit 1)

A'. 还原 bundle   → Serve fast-path bundle closure check passed.            (exit 0)
```

那一个静态 import 就把**整个约 179 KB 的 ACP runtime 簇**（bridge → compactionEngine、bridgeClient、permissionMediator……）拉进了一个静态可达 chunk —— 正是这个 guard 要拦的"重 runtime 通过 shared chunk 进入 listener 路径"的回归。基于 metafile 可达性的方法（跟随静态 `import`/`require` 边、跳过 `dynamic-import` + `external`）正确抓到了它，同时动态 import 路径仍被允许（由脚本单测覆盖）。

### 说明 / 取舍（非阻塞）
- **覆盖范围迁移且扩大。** 被删掉的 in-vitest 断言只检查 `run-qwen-serve` 这一个 root、4 个 forbidden 输入；新脚本检查全部 3 个 pre-listen root、7 个 ACP/core-shell runtime 模块，**外加** 4 个重 vendor 包（glob / chokidar / @iarna/toml / fzf），还带 import-path 诊断。覆盖严格更广。
- **仅在 ubuntu 跑**（`test` job 的静态检查腿），而旧断言在每个平台的 vitest 里都跑。bundle import 图与平台无关，所以 ubuntu-only 足够，并且从 3 个平台的单测里各去掉了一次约 30s 的 esbuild —— CI 净收益（PR 也说明了新增的 CLI-only build 成本）。
- **不再属于 `npm test`。** bundle 级回归现在由 `npm run check:serve-fast-path-bundle` / CI 抓，而非裸 `npm test`。`fast-path.test.ts` 里的快速源码级 guard 仍在单测里跑，所以源码级违规本地仍能抓到；只有 esbuild shared-chunk 的边角情况需要专门检查。这是有意且合理的取舍。
- 找不到 root output 时 guard 会 fail-closed（"Could not find bundled outputs …"）—— 安全默认（有测试覆盖）。

### 建议
**无需改动。** guard 名副其实、测试充分，import 边界在真实 bundle 上成立。建议在进行中的 CI 变绿后合并 —— 报告时新的 **"Check serve fast-path bundle closure"** 步骤还在 ubuntu 腿上运行（本地通过）；mac/windows 是预期的 named-job skip。

</details>

<sub>Verified on macOS against PR head `a6cc23cca0`: clean→PASS, static-import mutation→FAIL, restore→PASS; 12 script + 58 source tests + cli typecheck all green.</sub>

**Review by wenshao — DISMISSED** [review]

**Comment by wenshao** [comment]
@qwen-code /triage

File: scripts/check-serve-fast-path-bundle.js [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+    } else {
+      missingRoots.push(`${root.label} (${root.suffixes.join(' or ')})`);
+    }
+  }
+
+  if (missingRoots.length > 0) {
+    throw new Error(
+      'Could not find bundled outputs for serve pre-listen roots:\n' +
+        missingRoots.map((root) => `- ${root}`).join('\n') +
+        '\nRun DEV=true npm run bundle before this check.',
```
Comment (wenshao) [inline scripts/check-serve-fast-path-bundle.js]: **[Suggestion]** The error message here says `'Run DEV=true npm run bundle before this check.'`, which has two problems: (1) it drops `cross-env`, so it would fail on Windows CMD where `DEV=true` is not valid syntax; (2) it omits the preceding `npm run build -- --cli-only` step that the npm script chains. A developer following this guidance on Windows or without a prior build would hit a secondary failure.

The same issue exists at line 271 (`'Run \`npm run check:serve-fast-path-bundle\` to build one.'`) — that command itself needs the metafile to exist, creating a circular dependency.

```suggestion
        '\nRun `npm run build -- --cli-only && cross-env DEV=true npm run bundle` to produce the metafile.',
```

_— qwen3.7-max via Qwen Code /review_

File: scripts/check-serve-fast-path-bundle.js [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+    }
+  }
+
+  return offenders;
+
+  function addOffender(label, matchedInput, outputPath) {
+    const key = `${label}\0${outputPath}`;
+    if (seen.has(key)) return;
+    seen.add(key);
+    offenders.push({
```
Comment (wenshao) [inline scripts/check-serve-fast-path-bundle.js:243]: **[Suggestion]** The dedup key `${label}\0${outputPath}` omits `matchedInput`. If one output chunk bundles two different files matching the same forbidden label (e.g., `bridge.ts` and `bridgeClient.ts` both under "ACP bridge runtime"), only the first offender is reported — the second is silently dropped.

This makes debugging harder: fixing one input doesn't reveal the other until the next CI run.

```suggestion
    const key = `${label}\0${matchedInput}\0${outputPath}`;
```

_— qwen3.7-max via Qwen Code /review_

File: scripts/check-serve-fast-path-bundle.js [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+  metafilePath = DEFAULT_METAFILE_PATH,
+} = {}) {
+  if (!existsSync(metafilePath)) {
+    throw new Error(
+      `Missing esbuild metafile at ${metafilePath}. ` +
+        'Run `npm run check:serve-fast-path-bundle` to build one.',
+    );
+  }
+
+  const metafile = JSON.parse(readFileSync(metafilePath, 'utf8'));
```
Comment (wenshao) [inline scripts/check-serve-fast-path-bundle.js]: **[Suggestion]** `JSON.parse(readFileSync(...))` has no try/catch wrapping. If the metafile is corrupted or truncated (e.g., build killed mid-write, disk full, artifact upload/download corruption), this produces a raw `SyntaxError` with no context about which file failed or how to fix it.

Since `checkServeFastPathBundle` is exported and can be used programmatically, callers get an uncontextualized error.

```suggestion
  let metafile;
  try {
    metafile = JSON.parse(readFileSync(metafilePath, 'utf8'));
  } catch (e) {
    throw new Error(
      `Invalid esbuild metafile at ${metafilePath}: ${e instanceof Error ? e.message : String(e)}. Re-run the build.`,
    );
  }
```

_— qwen3.7-max via Qwen Code /review_

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -247,6 +247,10 @@ jobs:
           fi
           echo "Settings schema is up-to-date"
 
+      - name: 'Check serve fast-path bundle closure'
+        if: "${{ needs.classify_pr.outputs.skip_ci != 'true' }}"
+        run: 'npm run check:serve-fast-path-bundle'
```
Comment (wenshao) [inline .github/workflows/ci.yml:252]: **[Suggestion]** The removed test (`keeps bundled runQwenServe static imports free of ACP runtime modules`) ran on all three CI platforms (Ubuntu, macOS, Windows) via `npm run test:ci`. The replacement step is only added to the Ubuntu PR job (`test`). The `test_macos` and `test_windows` merge-queue jobs no longer verify the bundle closure invariant.

Risk is low since esbuild output is deterministic across platforms and the check script normalizes paths, but this is an unintentional regression in CI coverage scope. Consider adding the step to the merge-queue jobs for parity, or noting the intentional platform reduction in the PR description.

_— qwen3.7-max via Qwen Code /review_

**Review by wenshao — APPROVED** [review]

**Comment by wenshao** [comment]
@qwen-code /triage
